OpenRA.Mods.Common/Traits/OwnerLostAction.cs

```diff
@@ -34,14 +34,23 @@ public class OwnerLostActionInfo : ConditionalTraitInfo
 		public override object Create(ActorInitializer init) { return new OwnerLostAction(init, this); }
 	}
 
-	public class OwnerLostAction : ConditionalTrait<OwnerLostActionInfo>, INotifyOwnerLost
+	public class OwnerLostAction : ConditionalTrait<OwnerLostActionInfo>, INotifyOwnerLost, ITick
 	{
+		bool ownerLost;
+
 		public OwnerLostAction(ActorInitializer init, OwnerLostActionInfo info)
 			: base(info) { }
 
 		void INotifyOwnerLost.OnOwnerLost(Actor self)
 		{
-			if (IsTraitDisabled)
+			// OnOwnerLost might get called before the actor's activity runs,
+			// so we defer the actual action to ITick.Tick to avoid potential NREs in activity code.
+			ownerLost = true;
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (IsTraitDisabled || !ownerLost)
```

Comment:
Note: Maybe I just lack the creativity, but I fail to see how this trait being conditional as anything but a potential source of bugs and headaches.

In my opinion either you want an actor to do something when the player loses, or you don't, but making this toggleable on the fly via conditions is just asking for trouble, and we're not doing that in any of the official mods.

If anyone can point me to a downstream example that showcases why conditionality on this trait is needed, I'm honestly interested.

Reply:
It was used in Gen Alpha as part of Driver Killing code there, but lemme be honest it was a pretty bad code (owner was kept and External Conditions were used, while in original games with this logic vehicle became civilian, i had made it so if loses Driver killed unit become Neutral anyway instead of dying). Not sure if there are any better use cases.